Prevent PRs from merging if they contain breaking changes or new features with incorrect target label

josephperrott opened the pull request:
See individual commits.

AndrewKushnir commented on dev-infra/pr/merge/pull-request.ts:
```diff
@@ -138,3 +170,37 @@ async function fetchPullRequestFromGithub(
+ */
+function assertCorrectTargetForChanges(rawCommits: string[], label: TargetLabel) {
+  /** List of ParsedCommits for all of the commits in the pull request. */
+  const commits = rawCommits.map(m => parseCommitMessage(m));
```
```suggestion
  const commits = rawCommits.map(parseCommitMessage);
```

AndrewKushnir commented on dev-infra/pr/merge/pull-request.ts:
```diff
@@ -114,16 +122,40 @@ export async function loadAndValidatePullRequest(
+    nodes: [{
+      commit: {
+        status: {
+          state: graphQLTypes.oneOf(['FAILURE' as const, 'PENDING' as const, 'SUCCESS' as const]),
```
Can we say `as const` at the end?
```suggestion
          state: graphQLTypes.oneOf(['FAILURE', 'PENDING', 'SUCCESS'] as const),
```

josephperrott replied:
Ahh, didn't know that you could do that.  Done.

AndrewKushnir commented on dev-infra/pr/merge/pull-request.ts:
```diff
@@ -138,3 +170,37 @@ async function fetchPullRequestFromGithub(
+        throw PullRequestFailure.hasBreakingChanges(label);
+      }
+      // Check if any commits in the PR contains a commit type of "feat".
+      if (commits.some(commit => commit.type === 'feat')) {
+        throw PullRequestFailure.hasFeatureCommits(label);
+      }
```
Should we have an exclusion here for `feat(dev-infra) ...` commits (configurable)?

gkalpak replied:
+1
Also for `feat(docs-infra)` 🙏

josephperrott replied:
Added as requested.

AndrewKushnir reviewed (commented):
@josephperrott thanks for taking care of it so quickly 👍
I've just left a few minor comments/questions.

AndrewKushnir commented:
Btw, I believe these changes would also simplify adding a verification (in future PRs) that there is a breaking changes note in one of the commit messages, when "flag: breaking changes" is applied to a PR (https://github.com/angular/angular/issues/38776).

gkalpak commented on dev-infra/pr/merge/pull-request.ts:
```diff
@@ -6,9 +6,13 @@
 
-import * as Octokit from '@octokit/rest';
+import {params, types as graphQLTypes} from 'typed-graphqlify';
+import {parseCommitMessage} from '../../commit-message/parse';
+import {red, warn} from '../../utils/console';
```
These two imports belong in the second commit.

gkalpak commented on dev-infra/pr/merge/failures.ts:
```diff
@@ -6,6 +6,8 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {TargetLabel} from './config';
```
This change belongs in the second commit.

gkalpak commented on dev-infra/pr/merge/failures.ts:
```diff
@@ -72,4 +74,17 @@ export class PullRequestFailure {
+        'commits with the "feat" type.  New features can only be merged with the "target: minor" or ' +
+        '"target: major" label.';
+    return new this(message);
+  }
```
These changes belong in the second commit.

gkalpak reviewed (approved):
:+1:

josephperrott committed:
refactor(dev-infra): use graphql to query PRs in merge tooling

Migrate to use graqhql to query for PR information during the validation and
preperation portions of the pr merge tooling.

josephperrott committed:
feat(dev-infra): prevent merges for PRs with invalid breaking changes or commit types

Check commits in pull requests to ensure the pr can be merged into the target branch. Confirms
that prs targeting minor do not contain breaking changes, and prs targeting patch or lts do not
contain breaking changes or `feat` commits.

josephperrott committed:
feat(dev-infra): add support for `targetLabelExemptScopes` for merging

Add a property, `targetLabelExemptScopes`, to the merge configuration allowing certain
scopes to be exempted from requirements for features and breaking changes only included
in PRs targetting certain labels.

josephperrott committed:
build: add `dev-infra` and `docs-infra` as targetLabelExemptScopes

Add these scopes to allow for use of features and breaking changes for the
scopes regardless of target labels.

josephperrott commented:
@AndrewKushnir I can do that in a follow up PR, I don't want to expand the scope of this PR, but it definitely should be simple to do.

devversion commented on dev-infra/pr/merge/pull-request.ts:
```diff
@@ -114,16 +123,40 @@ export async function loadAndValidatePullRequest(
+const PR_SCHEMA = {
+  url: graphQLTypes.string,
+  number: graphQLTypes.number,
+  commits: params({first: 100}, {
```
It is unrealistic that there are PRs with more than 100 commits, but I think we should add a comment. We build up the `PullRequest` object with `commitCount` etc. but in practice this number can be wrong, right?

devversion replied:
Also for the sake of the "last commit status", should we get the last 100 instead?

josephperrott replied:
- updated to use the last 100 commits
- noted that we only look at the last 100 commits as pull requests with >100 commits are essentially never happening in our repos
- Used `totalCount` instead of the length of the `commits.nodes` array

devversion commented on dev-infra/pr/merge/pull-request.ts:
```diff
@@ -138,3 +171,45 @@ async function fetchPullRequestFromGithub(
+  switch (label.pattern) {
+    case 'target: major':
+      break;
+    case 'target: minor':
```
Maybe I'm overthinking this, but would it be more aligned with all the other code we have, if we just handled this inside the default target label configuration? (so that we don't need to hard-code label names here etc.).

We could just have a property like `validate` in the label configuration and pass it the list of `Commit`'s.

josephperrott replied:
I don't think that is overthinking, however I would actually propose that we remove the default target label configuration function, and instead allow configurations to extend from that initial set of target labels.

So then we would in all of our tooling assume that the repo will have support to use `target: {major,minor,patch,lts}`

devversion replied:
That makes sense to me. Though, I feel like it would be better to have logic such as the validation above, grouped together with the other label/branch configuration (hence my proposal to have something like `validate`).

Not too sure if it makes sense though. I'm just a little concerned about spreading such logic in multiple places.

josephperrott replied:
I think that having this `validate` function on a target label configuration ends up spreading out the logic even more.

Additionally, we would need to allow for the target label configuration functions to throw a `PullRequestFailure` which I think makes the most sense to keep local to this module.

devversion replied:
Yeah, I think you're right. We'll just have another place where the label name is hard-coded. Ideally we'd use an enum for this long-term. We can just do this in a follow-up.

josephperrott committed:
fixup! feat(dev-infra): add support for `targetLabelExemptScopes` for merging

devversion reviewed (approved):
LGTM
